### ISSUE TITLE
`iceberg-spark-runtime-3.3_2.12-1.5.1` seems to be compiled with a mismatched scala version

### ISSUE THREAD

**Issue by wForget** [apache/iceberg#10251]
### Apache Iceberg version

1.5.1 (latest release)

### Query engine

Spark

### Please describe the bug 🐞

As commented in https://github.com/apache/kyuubi/pull/6349#issuecomment-2082258210, `iceberg-spark-runtime-3.3_2.12-1.5.1` seems to be compiled with a mismatched scala version.

I decompiled the `org.apache.spark.sql.catalyst.plans.logical.DeleteFromIcebergTable` class in `iceberg-spark-runtime-3.3_2.12-1.5.1.jar` and found that it contains `scala/collection/SeqOps` class of scala 2.13.X.

Command:

```
javap -cp ./iceberg-spark-runtime-3.3_2.12-1.5.1.jar -v org.apache.spark.sql.catalyst.plans.logical.DeleteFromIcebergTable | grep 'scala/collection/SeqOps'
```

Result:

```
$ javap -cp ./iceberg-spark-runtime-3.3_2.12-1.5.1.jar -v org.apache.spark.sql.catalyst.plans.logical.DeleteFromIcebergTable | grep 'scala/collection/SeqOps'
  #221 = Utf8               (Lscala/collection/SeqOps;)Lscala/collection/SeqOps;
  #222 = NameAndType        #220:#221     // unapplySeq:(Lscala/collection/SeqOps;)Lscala/collection/SeqOps;
  #223 = Methodref          #219.#222     // scala/collection/immutable/Seq$.unapplySeq:(Lscala/collection/SeqOps;)Lscala/collection/SeqOps;
  #228 = Utf8               (Lscala/collection/SeqOps;)Z
  #229 = NameAndType        #227:#228     // isEmpty$extension:(Lscala/collection/SeqOps;)Z
  #230 = Methodref          #29.#229      // scala/collection/SeqFactory$UnapplySeqWrapper$.isEmpty$extension:(Lscala/collection/SeqOps;)Z
  #232 = NameAndType        #231:#221     // get$extension:(Lscala/collection/SeqOps;)Lscala/collection/SeqOps;
  #233 = Methodref          #29.#232      // scala/collection/SeqFactory$UnapplySeqWrapper$.get$extension:(Lscala/collection/SeqOps;)Lscala/collection/SeqOps;
  #234 = Utf8               (Lscala/collection/SeqOps;)V
  #235 = NameAndType        #141:#234     // "<init>":(Lscala/collection/SeqOps;)V
  #236 = Methodref          #24.#235      // scala/collection/SeqFactory$UnapplySeqWrapper."<init>":(Lscala/collection/SeqOps;)V
  #238 = Utf8               (Lscala/collection/SeqOps;I)I
  #239 = NameAndType        #237:#238     // lengthCompare$extension:(Lscala/collection/SeqOps;I)I
  #240 = Methodref          #29.#239      // scala/collection/SeqFactory$UnapplySeqWrapper$.lengthCompare$extension:(Lscala/collection/SeqOps;I)I
  #242 = Utf8               (Lscala/collection/SeqOps;I)Ljava/lang/Object;
  #243 = NameAndType        #241:#242     // apply$extension:(Lscala/collection/SeqOps;I)Ljava/lang/Object;
  #244 = Methodref          #29.#243      // scala/collection/SeqFactory$UnapplySeqWrapper$.apply$extension:(Lscala/collection/SeqOps;I)Ljava/lang/Object;
  #250 = Utf8               scala/collection/SeqOps
  #251 = Class              #250          // scala/collection/SeqOps
        92: invokevirtual #223                // Method scala/collection/immutable/Seq$.unapplySeq:(Lscala/collection/SeqOps;)Lscala/collection/SeqOps;
       102: invokevirtual #230                // Method scala/collection/SeqFactory$UnapplySeqWrapper$.isEmpty$extension:(Lscala/collection/SeqOps;)Z
       117: invokevirtual #233                // Method scala/collection/SeqFactory$UnapplySeqWrapper$.get$extension:(Lscala/collection/SeqOps;)Lscala/collection/SeqOps;
       120: invokespecial #236                // Method scala/collection/SeqFactory$UnapplySeqWrapper."<init>":(Lscala/collection/SeqOps;)V
       134: invokevirtual #233                // Method scala/collection/SeqFactory$UnapplySeqWrapper$.get$extension:(Lscala/collection/SeqOps;)Lscala/collection/SeqOps;
       138: invokevirtual #240                // Method scala/collection/SeqFactory$UnapplySeqWrapper$.lengthCompare$extension:(Lscala/collection/SeqOps;I)I
       153: invokevirtual #233                // Method scala/collection/SeqFactory$UnapplySeqWrapper$.get$extension:(Lscala/collection/SeqOps;)Lscala/collection/SeqOps;
       157: invokevirtual #244                // Method scala/collection/SeqFactory$UnapplySeqWrapper$.apply$extension:(Lscala/collection/SeqOps;I)Ljava/lang/Object;
       173: invokevirtual #233                // Method scala/collection/SeqFactory$UnapplySeqWrapper$.get$extension:(Lscala/collection/SeqOps;)Lscala/collection/SeqOps;
       177: invokevirtual #244                // Method scala/collection/SeqFactory$UnapplySeqWrapper$.apply$extension:(Lscala/collection/SeqOps;I)Ljava/lang/Object;
          locals = [ class org/apache/spark/sql/catalyst/plans/logical/DeleteFromIcebergTable, class scala/collection/immutable/IndexedSeq, top, top, top, top, top, class scala/collection/immutable/IndexedSeq, class scala/collection/SeqOps ]
          locals = [ class org/apache/spark/sql/catalyst/plans/logical/DeleteFromIcebergTable, class scala/collection/immutable/IndexedSeq, top, top, top, top, top, class scala/collection/immutable/IndexedSeq, class scala/collection/SeqOps, class org/apache/spark/sql/catalyst/plans/logical/LogicalPlan, class org/apache/spark/sql/catalyst/plans/logical/LogicalPlan ]
```

**Comment by wForget** [comment]
cc @Fokko

**Comment by pan3793** [comment]
cc 1.5.1 release manager @amogh-jahagirdar

**Comment by nastra** [comment]
I just confirmed that this is an issue with the Spark 3.3 artifacts for Iceberg 1.5.1. The same artifacts for Iceberg 1.5.0 are correct and work fine.

**Comment by wForget** [comment]
> I just confirmed that this is an issue with the Spark 3.3 artifacts for Iceberg 1.5.1. The same artifacts for Iceberg 1.5.0 are correct and work fine.

Does `iceberg-spark-runtime-3.5_2.13-1.5.1` have similar problem?

**Comment by ajantha-bhat** [comment]
we found similar issue while bumping the jars at Nessie side aswell.
https://github.com/projectnessie/nessie/pull/8403

**Comment by amogh-jahagirdar** [comment]
Taking a look at all this, it seems like something messed up when I ran the `stage-binaries.sh` script as part of the release. I remember there was an intermittent connection issue and I restarted the whole script but I figured that should be fine to do since we always use an explicit scala version at each stage in the script?  https://github.com/apache/iceberg/blob/main/dev/stage-binaries.sh

**Comment by amogh-jahagirdar** [comment]
I'm looking into next steps here, if we need to do another release or since there is no really source change that we need to do and rather about fixing the released artifacts if there's some other way forward.

**Comment by pan3793** [comment]
Maven Central does not allow overwriting artifacts, consider starting the next release soon?

**Comment by Fokko** [comment]
Thanks for raising this @wForget 

I'm unsure how this happened since I explicitly checked against the Spark 3.5 Scala 2.12 jars https://github.com/tabular-io/docker-spark-iceberg/pull/149/commits/8bb375cdbce4cf6d22b39a7345570a5fed1b150f

**Comment by amogh-jahagirdar** [comment]
It's a strange mix of artifacts that are incorrect. 

These are the ones I've that I've seen so far that are incorrect: 
1.) Spark3.5+Scala2.13
2.) Spark3.3+Scala2.12

The ones that are OK from my testing so far:
1.) Spark3.5 + Scala 2.12
2.) Spark3.4 + Scala 2.12

**Comment by wForget** [comment]
> Taking a look at all this, it seems like something messed up when I ran the `stage-binaries.sh` script as part of the release. I remember there was an intermittent connection issue and I restarted the whole script but I figured that should be fine to do since we always use an explicit scala version at each stage in the script? https://github.com/apache/iceberg/blob/main/dev/stage-binaries.sh

Is it missing `clean` task? (I'm not familiar with gradle)

**Comment by ajantha-bhat** [comment]
@wForget, @kapkiai, @pan3793 : 
Feel free to test the Iceberg 1.5.2 staged artifacts  

I just tested with Nessie and it can work fine now (https://github.com/projectnessie/nessie/pull/8444)

**Comment by wForget** [comment]
> Feel free to test the Iceberg 1.5.2 staged artifacts

Kyuubi CI with iceberg 1.5.2 RC0 also works fine. https://github.com/apache/kyuubi/pull/6361

**Comment by pan3793** [comment]
Followup of @wForget comments: Kyuubi's CI covers Iceberg
- Spark 3.3-3.5 with Scala 2.12
- Spark 3.5 with Scala 2.13

**Comment by Fokko** [comment]
Thanks everyone for testing! Make sure to share your findings on the devlist: https://lists.apache.org/thread/j6kszdofsbxyv7cckmkd898hn2zfngpr

**Comment by nastra** [comment]
I'll go ahead and close this, since this was fixed by releasing Iceberg 1.5.2 with the correctly compiled artifacts.
